File: 2-Graph Search, Shortest Path, and Data Structures/3-Graph Search (BFS & DFS)/Python/undirected_graph.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Adjacency list representation of undirected graph.

Note that parallel edges and self-loops are not allowed.
"""

__author__ = 'Ziang Lu'

from collections import deque

from graph_basics import AbstractGraph, AbstractVertex


class IllegalArgumentError(ValueError):
    pass


class Vertex(AbstractVertex):
    __slots__ = ['_edges', '_neighbors']

    def __init__(self, vtx_id: int):
        """
        Constructor with parameter.
        :param vtx_id: int
        """
        super().__init__(vtx_id)
        self._edges = []
        self._neighbors = set()

    def get_edge_with_neighbor(self, neighbor: AbstractVertex):
        """
        Returns the first edge with the given neighbor.
        :param neighbor: AbstractVertex
        :return: UndirectedEdge
        """
        # Check whether the input neighbor is None
        if not neighbor:
            raise IllegalArgumentError('The input neighbor should not be None.')

        for edge in self._edges:
            if (edge.end1 is self and edge.end2 is neighbor) or \
                    (edge.end1 is neighbor and edge.end2 is self):
                return edge
        # Not found
        return None

    @property
    def edges(self) -> list:
        """
        Accessor of edges.
        :return: list[UndirectedEdge]
        """
        return self._edges

    def add_edge(self, new_edge) -> None:
        """
        Adds the given edge to this vertex.
        :param new_edge: UndirectedEdge
        :return: None
        """
        # Check whether the input edge is None
        if not new_edge:
            raise IllegalArgumentError('The edge to add should not be None.')
        # Check whether the input edge involves this vertex
        if new_edge.end1 is not self and new_edge.end2 is not self:
            raise IllegalArgumentError(
                'The edge to add should involve this vertex.'
            )
        # Find the neighbor associated with the input edge
        if new_edge.end1 is self:  # endpoint2 is the neighbor.
            neighbor = new_edge.end2
        else:  # endpoint1 is the neighbor.
            neighbor = new_edge.end1
        # Check whether the input edge already exists
        if neighbor.vtx_id in self._neighbors:
            raise IllegalArgumentError('The edge to add already exists.')

        self._edges.append(new_edge)
        self._neighbors.add(neighbor.vtx_id)

    def remove_edge(self, edge_to_remove) -> None:
        """
        Removes the given edge from this vertex.
        :param edge_to_remove: UndirectedEdge
        :return: None
        """
        # Check whether the input edge is None
        if not edge_to_remove:
            raise IllegalArgumentError('The edge to remove should not be None.')
        # Check whether the input edge involves this vertex
        if (edge_to_remove.end1 is not self) and \
                (edge_to_remove.end2 is not self):
            raise IllegalArgumentError(
                'The edge to remove should involve this vertex.'
            )
        # Find the neighbor associated with the input edge
        if edge_to_remove.end1 is self:  # endpoint2 is the neighbor.
            neighbor = edge_to_remove.end2
        else:  # endpoint1 is the neighbor.
            neighbor = edge_to_remove.end1
        # Check whether the input edge exists
        if neighbor.vtx_id not in self._neighbors:
            raise IllegalArgumentError("The edge to remove doesn't exist.")

        self._edges.remove(edge_to_remove)
        self._neighbors.remove(neighbor.vtx_id)

    def __repr__(self):
        return f'Vertex #{self._vtx_id}, Its neighbors: {self._neighbors}'


class UndirectedEdge(object):
    __slots__ = ['_end1', '_end2']

    def __init__(self, end1: Vertex, end2: Vertex):
        """
        Constructor with parameter.
        :param end1: Vertex
        :param end2: Vertex
        """
        self._end1 = end1
        self._end2 = end2

    @property
    def end1(self) -> Vertex:
        """
        Accessor of end1.
        :return: Vertex
        """
        return self._end1

    @property
    def end2(self) -> Vertex:
        """
        Accessor of end2.
        :return: Vertex
        """
        return self._end2

    @end1.setter
    def end1(self, end1: Vertex) -> None:
        """
        Mutator of end1.
        :param end1: Vertex
        :return: None
        """
        self._end1 = end1

    @end2.setter
    def end2(self, end2: Vertex) -> None:
        """
        Mutator of end2.
        :param end2: Vertex
        :return: None
        """
        self._end2 = end2

    def __repr__(self):
        return f'Edge between Vertex #{self._end1.vtx_id} and Vertex #{self._end2.vtx_id}'


class UndirectedGraph(AbstractGraph):
    __slots__ = []

    def __init__(self):
        """
        Default constructor.
        """
        super().__init__()

    def add_vtx(self, new_vtx_id):
        # Check whether the input vertex is repeated
        if self._find_vtx(new_vtx_id):
            raise IllegalArgumentError('The input vertex is repeated.')

        new_vtx = Vertex(new_vtx_id)
        self._vtx_list.append(new_vtx)

    def _remove_vtx(self, vtx_to_remove):
        # Remove all the edges associated with the vertex to remove
        edges_to_remove = vtx_to_remove.edges
        while len(edges_to_remove):
            self._remove_edge(edge_to_remove=edges_to_remove[0])
        # Remove the vertex
        self._vtx_list.remove(vtx_to_remove)

    def add_edge(self, end1_id, end2_id):
        # Check whether the input endpoints both exist
        end1, end2 = self._find_vtx(end1_id), self._find_vtx(end2_id)
        if not end1 or not end2:
            raise IllegalArgumentError("The endpoints don't both exist.")
        # Check whether the input endpoints are the same (self-loop)
        if end1_id == end2_id:
            raise IllegalArgumentError(
                'The endpoints are the same (self-loop).'
            )

        new_edge = UndirectedEdge(end1, end2)
        self._add_edge(new_edge=new_edge)

    def _add_edge(self, new_edge):
        end1, end2 = new_edge.end1, new_edge.end2
        end1.add_edge(new_edge)
        end2.add_edge(new_edge)
        self._edge_list.append(new_edge)

    def remove_edge(self, end1_id, end2_id):
        # Check whether the input endpoints both exist
        end1, end2 = self._find_vtx(end1_id), self._find_vtx(vtx_id=end2_id)
        if not end1 or not end2:
            raise IllegalArgumentError("The endpoints don't both exist.")

        # Check whether the edge to remove exists
        edge_to_remove = end1.get_edge_with_neighbor(end2)
        if not edge_to_remove:
            raise IllegalArgumentError("The edge to remove doesn't exist.")

        self._remove_edge(edge_to_remove=edge_to_remove)

    def _remove_edge(self, edge_to_remove):
        end1, end2 = edge_to_remove.end1, edge_to_remove.end2
        end1.remove_edge(edge_to_remove)
        end2.remove_edge(edge_to_remove)
        self._edge_list.remove(edge_to_remove)

    def bfs(self, src_vtx_id):
        # Check whether the input source vertex exists
        src_vtx = self._find_vtx(src_vtx_id)
        if not src_vtx:
            raise IllegalArgumentError("The input source vertex doesn't exist.")

        # 1. Initialize G as s explored and other vertices unexplored
        src_vtx.set_as_explored()
        # 2. Let Q be the queue of vertices initialized with s
        queue = deque()
        queue.append(src_vtx)

        findable_vtx_ids = [src_vtx_id]

        # 3. While Q is not empty
        while queue:
            # (1) Take out the first vertex v
            vtx = queue.popleft()
            # (2) For every edge (v, w)
            for edge in vtx.edges:
                # Find the neighbor
                if edge.end1 is vtx:  # endpoint2 is the neighbor.
                    neighbor = edge.end2
                else:  # endpoint1 is the neighbor.
                    neighbor = edge.end1
                # If w is unexplored
                if not neighbor.explored:
                    # Mark w as explored
                    neighbor.set_as_explored()

                    findable_vtx_ids.append(neighbor.vtx_id)

                    # Push w to Q
                    queue.append(neighbor)
        return findable_vtx_ids

    def shortest_path(self, src_vtx_id, dest_vtx_id):
        # Check whether the input source and destination vertices both exist
        src_vtx, dest_vtx = self._find_vtx(src_vtx_id), \
            self._find_vtx(dest_vtx_id)
        if not src_vtx:
            raise IllegalArgumentError("The input source and destination "
                                       "vertices don't both exist.")

        # 1. Initialize G as s explored and other vertices unexplored
        src_vtx.set_as_explored()
        # 2. Let Q be the queue of vertices initialized with s
        queue = deque()
        queue.append(src_vtx)
        # 3. While Q is not empty
        while queue:
            # (1) Take out the first vertex v
            vtx = queue.popleft()
            # (2) For every edge (v, w)
            for edge in vtx.edges:
                # Find the neighbor
                if edge.end1 is vtx:  # endpoint2 is the neighbor.
                    neighbor = edge.end2
                else:  # endpoint1 is the neighbor.
                    neighbor = edge.end1
                # If w is unexplored
                if not neighbor.explored:
                    # Mark w as explored
                    neighbor.set_as_explored()

                    neighbor.layer = vtx.layer + 1
                    if neighbor is dest_vtx:  # Found it
                        return dest_vtx.layer

                    # Push w to Q
                    queue.append(neighbor)
        # The destination vertex is not findable starting from the given source
        # vertex.
        return -1

    def num_of_connected_components_with_bfs(self):
        # Undirected connectivity
        components = []
        for vtx in self._vtx_list:
            # If v is unexplored (i.e., not explored from some previous BFS)
            if not vtx.explored:
                # Do BFS towards v
                # (Discovers precisely v's connected component)
                component = self.bfs(vtx.vtx_id)
                components.append(component)
        return len(components)

    def _dfs_helper(self, vtx, findable_vtx_ids):
        # For every edge (v, w)
        for edge in vtx.edges:
            # Find the neighbor
            if edge.end1 is vtx:  # endpoint2 is the neighbor.
                neighbor = edge.end2
            else:  # endpoint1 is the neighbor.
                neighbor = edge.end1
            # If w is unexplored
            # (This itself serves as a base case: all the w's of v are
            # explored.)
            if not neighbor.explored:
                # Mark w as explored
                neighbor.set_as_explored()

                findable_vtx_ids.append(neighbor.vtx_id)

                # Do DFS on (G, w)   (Recursion)
                self._dfs_helper(neighbor, findable_vtx_ids=findable_vtx_ids)

    def num_of_connected_components_with_dfs(self):
        # Undirected connectivity
        components = []
        # For every vertex v
        for vtx in self._vtx_list:
            # If v is unexplored (i.e., not explored from some previous DFS)
            if not vtx.explored:
                # Do DFS towards v
                # (Discovers precisely v's connected component)
                component = self.dfs(vtx.vtx_id)
                components.append(component)
        return len(components)

File: 3-Greedy Algorithms, Minimum Spanning Tree, and Dynamic Programming/1-Greedy Algorithm/3-Minimum Spanning Tree/Java/undirected_graph/UndirectedGraph.java
package undirected_graph;

import java.util.ArrayList;
import java.util.BitSet;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Random;

import graph.GraphInterface;
import union_find.UnionFind;

/**
 * Adjacency list representation of a undirected graph.
 *
 * Note that parallel edges are allowed, but not self-loops.
 * @author Ziang Lu
 */
public class UndirectedGraph implements GraphInterface {

    /**
     * Vertex list.
     */
    private final List<Vertex> vtxList;
    /**
     * Edge list.
     */
    private final List<UndirectedEdge> edgeList;

    /**
     * Default constructor.
     */
    public UndirectedGraph() {
        vtxList = new ArrayList<>();
        edgeList = new ArrayList<>();
    }

    @Override
    public void addVtx(int newVtxID) {
        // Check whether the input vertex is repeated
        if (findVtx(newVtxID) != null) {
            throw new IllegalArgumentException("The input vertex is repeated.");
        }

        Vertex newVtx = new Vertex(newVtxID);
        vtxList.add(newVtx);
    }

    /**
     * Private helper method to find the given vertex in this adjacency list.
     * @param vtxID vertex ID to look for
     * @return vertex if found, null if not found
     */
    private Vertex findVtx(int vtxID) {
        for (Vertex vtx : vtxList) {
            if (vtx.id() == vtxID) {
                return vtx;
            }
        }
        // Not found
        return null;
    }

    @Override
    public void removeVtx(int vtxID) {
        // Check whether the input vertex exists
        Vertex vtxToRemove = findVtx(vtxID);
        if (vtxToRemove == null) {
            throw new IllegalArgumentException("The input vertex doesn't exist.");
        }

        removeVtx(vtxToRemove);
    }

    /**
     * Private helper method to remove the given vertex from this graph.
     * @param vtxToRemove vertex to remove
     */
    private void removeVtx(Vertex vtxToRemove) {
        // Remove all the edges associated with the vertex to remove
        List<UndirectedEdge> edgesToRemove = vtxToRemove.edges();
        while (edgesToRemove.size() > 0) {
            removeEdge(edgesToRemove.get(0));
        }
        // Remove the vertex
        vtxList.remove(vtxToRemove);
    }

    @Override
    public void addEdge(int end1ID, int end2ID, double length) {
        // Check whether the input endpoints both exist
        Vertex end1 = findVtx(end1ID), end2 = findVtx(end2ID);
        if ((end1 == null) || (end2 == null)) {
            throw new IllegalArgumentException("The endpoints don't both exist.");
        }
        // Check whether the input endpoints are the same (self-loop)
        if (end1ID == end2ID) {
            throw new IllegalArgumentException("The endpoints are the same (self-loop).");
        }

        UndirectedEdge newEdge = new UndirectedEdge(end1, end2, length);
        addEdge(newEdge);
    }

    /**
     * Private helper method to add the given edge to this graph.
     * @param newEdge new edge
     */
    private void addEdge(UndirectedEdge newEdge) {
        Vertex end1 = newEdge.end1(), end2 = newEdge.end2();
        end1.addEdge(newEdge);
        end2.addEdge(newEdge);
        edgeList.add(newEdge);
    }

    @Override
    public void removeEdge(int end1ID, int end2ID) {
        // Check whether the input vertices both exist
        Vertex end1 = findVtx(end1ID), end2 = findVtx(end2ID);
        if ((end1 == null) || (end2 == null)) {
            throw new IllegalArgumentException("The input vertices don't both exist.");
        }
        // Check whether the edge to remove exists
        UndirectedEdge edgeToRemove = end1.getEdgeWithNeighbor(end2);
        if (edgeToRemove == null) {
            throw new IllegalArgumentException("The edge to remove doesn't exist.");
        }

        removeEdge(edgeToRemove);
    }

    /**
     * Private helper method to remove the given edge from this graph.
     * @param edgeToRemove edge to remove
     */
    private void removeEdge(UndirectedEdge edgeToRemove) {
        Vertex end1 = edgeToRemove.end1(), end2 = edgeToRemove.end2();
        end1.removeEdge(edgeToRemove);
        end2.removeEdge(edgeToRemove);
        edgeList.remove(edgeToRemove);
    }

    /**
     * Removes all the edges between a vertex pair from this graph.
     * @param end1ID endpoint1 ID
     * @param end2ID endpoint2 ID
     */
    public void removeEdgesBetweenPair(int end1ID, int end2ID) {
        try {
            while (true) {
                removeEdge(end1ID, end2ID);
            }
        } catch (IllegalArgumentException ex) {}
    }

    @Override
    public void showGraph() {
        System.out.println("The vertices are:");
        for (Vertex vtx : vtxList) {
            System.out.println(vtx);
        }
        System.out.println("The edges are:");
        for (UndirectedEdge edge : edgeList) {
            System.out.println(edge);
        }
    }

    /**
     * Finds the minimum spanning tree (MST) in this graph using straightforward
     * Prim's MST Algorithm.
     * @return cost of the MST
     */
    public double primMSTStraightforward() {
        // 1. Arbitrarily choose a source vertex s
        Random randomGenerator = new Random();
        Vertex srcVtx = vtxList.get(randomGenerator.nextInt(vtxList.size()));

        // 2. Initialize X = {s}, which contains the vertices we've spanned so far, and T = {empty}, which is the
        //    current spanning tree
        BitSet spanned = new BitSet();
        spanned.set(srcVtx.id());
        List<UndirectedEdge> currSpanningTree = new ArrayList<>();

        // 3. Create a heap containing all the edges with one endpoint in X and the other in (V-X)
        PriorityQueue<UndirectedEdge> crossingEdges = new PriorityQueue<>(srcVtx.edges());

        // 4. While X != V
        while (spanned.cardinality() < vtxList.size()) {
            // Among all crossing edges e = (v, w) with v in X and w in (V-X), pick up the cheapest crossing edge
            UndirectedEdge cheapestCrossingEdge = crossingEdges.poll();
            // Add e to T
            currSpanningTree.add(cheapestCrossingEdge);
            // Add w to X
            Vertex w = null;
            if (spanned.get(cheapestCrossingEdge.end1().id())) { // endpoint2 is the w.
                w = cheapestCrossingEdge.end2();
            } else { // endpoint1 is the w.
                w = cheapestCrossingEdge.end1();
            }
            spanned.set(w.id());

            // Update the crossing edges with w's edges if necessary
            for (UndirectedEdge wEdge : w.edges()) {
                // Find the neighbor
                Vertex neighbor = null;
                if (wEdge.end1().id() == w.id()) { // endpoint2 is the neighbor.
                    neighbor = wEdge.end2();
                } else { // endpoint1 is the neighbor.
                    neighbor = wEdge.end1();
                }
                // Check whether the neighbor of w has been spanned
                if (!spanned.get(neighbor.id())) {
                    crossingEdges.offer(wEdge);
                }
            }
        }

        return currSpanningTree.stream().mapToDouble(edge -> edge.cost()).sum();
        // Overall running time complexity: O((m + n)log m)
        // Since usually m >= n, it could be simplified to O(mlog m).
    }

    /**
     * Finds the minimum spanning tree (MST) in this graph using improved Prim's
     * MST Algorithm.
     * The improvement comes from the idea of Dijkstra's Shortest-Path
     * Algorithm, especially the invariants of the algorithm and the proper
     * usage of the heap.
     * @return cost of the MST
     */
    public double primMSTImproved() {
        // 1. Arbitrarily choose a source vertex s
        Random randomGenerator = new Random();
        Vertex srcVtx = vtxList.get(randomGenerator.nextInt(vtxList.size()));

        // 2. Initialize X = {s}, which contains the vertices we've spanned so far, and T = {empty}, which is the
        //    current spanning tree
        BitSet spanned = new BitSet();
        spanned.set(srcVtx.id());
        for (UndirectedEdge edge : srcVtx.edges()) {
            // Find the neighbor
            Vertex neighbor = null;
            if (edge.end1().id() == srcVtx.id()) { // endpoint2 is the neighbor.
                neighbor = edge.end2();
            } else { // endpoint1 is the neighbor.
                neighbor = edge.end1();
            }
            neighbor.setMinCostIncidentEdge(edge);
            neighbor.setMinIncidentCost(edge.cost());
        }
        List<UndirectedEdge> currSpanningTree = new ArrayList<>();

        // 3. Create a heap containing all the vertices not in X (V-X)
        PriorityQueue<Vertex> vtxsToProcess = new PriorityQueue<>(vtxList);

        // 4. While X != V
        while (spanned.cardinality() < vtxList.size()) {
            // Among all crossing edges e = (v, w) with v in X and w in (V-X) (the heap), pick up the cheapest crossing
            // edge
            Vertex w = vtxsToProcess.poll();
            // Add e to T
            currSpanningTree.add(w.minCostIncidentEdge());
            // Add w to X
            spanned.set(w.id());

            // Extracting one vertex from the heap (V-X) may influence some minimum cost of the incident edges from X
            // of the vertices that are still in the heap (V-X).
            // Vertices that are not connected to w and are still in the heap (V-X) won't be influenced.
            // => The minimum cost of the incident edges from X of vertices that are connected to w and are still in the
            //    heap (V-X) may drop down.

            // Update the minimum cost of the incident edges from X for the vertices if necessary
            for (UndirectedEdge wEdge : w.edges()) {
                // Find the neighbor
                Vertex neighbor = null;
                if (wEdge.end1().id() == w.id()) { // endpoint2 is the neighbor.
                    neighbor = wEdge.end2();
                } else { // endpoint1 is the neighbor.
                    neighbor = wEdge.end1();
                }
                // Check whether the neighbor of w has been spanned
                if (!spanned.get(neighbor.id())) {
                    // Check whether the minimum cost of the incident edges from X needs to be updated
                    double newCost = wEdge.cost();
                    if (newCost < neighbor.minIncidentCost()) {
                        // Remove this neighbor from the heap (V-X)
                        vtxsToProcess.remove(neighbor);
                        // Update its minimum cost of the incident edges from X
                        neighbor.setMinCostIncidentEdge(wEdge);
                        neighbor.setMinIncidentCost(newCost);
                        // Put this neighbor back to the heap (V-X)
                        vtxsToProcess.offer(neighbor);
                    }
                }
            }
        }

        return currSpanningTree.stream().mapToDouble(edge -> edge.cost()).sum();
        // Overall running time complexity: O((m + n)log n)
        // Since usually m >= n, it could be simplified to O(mlog n).
    }

    /**
     * Finds the minimum spanning tree (MST) in this graph using straightforward
     * Kruskal's MST Algorithm.
     * @return cost of the MST
     */
    public double kruskalMSTStraightforward() {
        // 1. Sort the edges in order of increasing cost   [O(mlog m)]
        List<UndirectedEdge> edges = new ArrayList<>(edgeList);
        Collections.sort(edges);

        // 2. Initialize T = {empty}, which is the current spanning tree
        List<UndirectedEdge> currSpanningTree = new ArrayList<>();

        // 3. For each edge e = (v, w) in the sorted edge list   [O(mn)]
        for (UndirectedEdge edge : edges) {
            // Check whether adding e to T causes cycles in T
            // This is equivalent to checking whether there exists a v-w path in T before adding e.
            if (!dfsAndCheckPath(currSpanningTree, edge.end1(), edge.end2())) {
                currSpanningTree.add(edge);
            }
        }

        return currSpanningTree.stream().mapToDouble(edge -> edge.cost()).sum();
        // Overall running time complexity: O(mn)
    }

    /**
     * Private helper method to check whether there exists a v-w path in the
     * given spanning tree.
     * @param spanningTree given spanning tree
     * @param v vertex v
     * @param w vertex w
     * @return whether v-w path exists in the given spanning tree
     */
    private boolean dfsAndCheckPath(List<UndirectedEdge> spanningTree, Vertex v, Vertex w) {
        // Create a map between vertices and its neighbors
        Map<Integer, List<Vertex>> connections = constructConnections(spanningTree);
        if (!connections.containsKey(v.id()) || !connections.containsKey(w.id())) {
            return false;
        }
        return dfsAndCheckPathHelper(connections, v, w);
        // Running time complexity: O(n)
    }

    /**
     * Helper method to construct the connection map from the given edges.
     * @param edges given edges
     * @return constructed connected map
     */
    private Map<Integer, List<Vertex>> constructConnections(List<UndirectedEdge> edges) {
        Map<Integer, List<Vertex>> connections = new HashMap<>();
        for (UndirectedEdge edge : edges) {
            addNeighbor(connections, edge.end1(), edge.end2());
            addNeighbor(connections, edge.end2(), edge.end1());
        }
        return connections;
        // Running time complexity: O(n)
    }

    /**
     * Helper method to add the given neighbor of the given vertex to the given
     * connection map.
     * @param connections given connection map
     * @param v given vertex
     * @param neighbor given neighbor
     */
    private void addNeighbor(Map<Integer, List<Vertex>> connections, Vertex v, Vertex neighbor) {
        List<Vertex> neighbors = connections.getOrDefault(v.id(), new ArrayList<>());
        neighbors.add(neighbor);
        connections.put(v.id(), neighbors);
        // Running time complexity: O(1)
    }

    /**
     * Helper method to check whether there exists a curr-target path in the
     * given connection map recursively.
     * @param spanningTree given spanning tree
     * @param curr current vertex
     * @param target target vertex
     * @return whether curr-target path exists in the given spanning tree
     */
    private boolean dfsAndCheckPathHelper(Map<Integer, List<Vertex>> connections, Vertex curr, Vertex target) {
        curr.setAsExplored();
        for (Vertex neighbor : connections.get(curr.id())) {
            if (neighbor.id() == target.id()) {
                return true;
            }
            if (!neighbor.explored()) {
                if (dfsAndCheckPathHelper(connections, neighbor, target)) {
                    return true;
                }
            }
        }
        return false;
        // Running time complexity: O(n)
    }

    /**
     * Finds the minimum spanning tree (MST) in this graph using improved
     * Kruskal's MST Algorithm.
     * @return cost of the MST
     */
    public double kruskalMSTImproved() {
        // 1. Sort the edges in order of increasing cost   [O(mlog m)]
        List<UndirectedEdge> edges = new ArrayList<>(edgeList);
        Collections.sort(edges);

        // 2. Initialize T = {empty}, which is the current spanning tree
        List<UndirectedEdge> currSpanningTree = new ArrayList<>();

        // 3. Create a UnionFind of vertices
        // object -> vertex
        // group -> connected component w.r.t. the edges in T
        // Each of the vertex is on its own isolated connected component.
        UnionFind<Vertex> unionFind = new UnionFind<>(vtxList);

        // 4. For each edge e = (v, w) in the sorted edge list   [O(nlog n)]
        for (UndirectedEdge edge : edges) {
            // Check whether adding e to T causes cycles in T
            // This is equivalent to checking whether there exists a v-w path in T before adding e.
            // This is equivalent to checking whether the leaders of v and w in the UnionFind are the same.
            if (edge.end1().leader() != edge.end2().leader()) {
                currSpanningTree.add(edge);
                // Fuse the two connected components to a single one
                String groupNameV = edge.end1().leader().objName(), groupNameW = edge.end2().leader().objName();
                unionFind.union(groupNameV, groupNameW);
            }
        }
        /*
         * Originally we would think it involves O(mn) leader updates; however,
         * we can change to a "vertex-centric" view:
         * Consider the number of leader updates for a single vertex:
         * Every time the leader of this vertex gets updated, the size of its
         * connected components at least doubles, so suppose it experiences x
         * leader updates in total, we have
         *     2^x <= n
         *     x <= log2 n
         * Thus, each vertex experiences O(log n) leader updates, leading to a
         * O(nlog n) leader updates in total.
         */

        return currSpanningTree.stream().mapToDouble(edge -> edge.cost()).sum();
        // Overall running time complexity: O(mlog m)
    }

    /**
     * Clusters the graph into the given number of clusters using maximum
     * spacing as the objective function, which is to maximize the minimum
     * distance between a pair of separated points, using Single-link Algorithm,
     * which is exactly the same as Kruskal's MST algorithm.
     * @param k number of clusters
     * @return maximum spacing of the clustering
     */
    public double clusteringWithMaxSpacing(int k) {
        // Check whether the input k is greater than 1
        if (k <= 1) {
            throw new IllegalArgumentException("The number of clusters must be at least 2.");
        }

        List<UndirectedEdge> edges = new ArrayList<>(edgeList);
        Collections.sort(edges);

        // Initially, each point is in a separate cluster.
        UnionFind<Vertex> unionFind = new UnionFind<>(vtxList);

        boolean stopped = false;
        for (UndirectedEdge edge : edges) {
            if (edge.end1().leader() != edge.end2().leader()) {
                if (stopped) {
                    return edge.cost();
                }
                // Let p, q = closest pair of separated points, which determines the current spacing
                // Merge the clusters containing p and q into a single cluster
                String groupNameP = edge.end1().leader().objName(), groupNameQ = edge.end2().leader().objName();
                unionFind.union(groupNameP, groupNameQ);
                if (unionFind.numOfGroups() == k) { // Repeat until only k clusters
                    // The maximum spacing is simply the cost of the next cheapest crossing edge among different
                    // connected components.
                    stopped = true;
                }
            }
        }
        return 0.0; // Codes should never reach here.
        // Overall running time complexity: O(mlog m)
    }

}
